### PR TITLE
Fix decrement metric when resource is deleted

### PR DESCRIPTION
**Pull request by sknot-rh**
Signed-off-by: Stanislav Knot <sknot@redhat.com>

### Type of change
- Bugfix

### Description
If all namespaces are watched, one verticle is used so we can reset all the records in the metrics map. If multiple namespaces are used, each of them has to reset its own part of the metrics map.

Fixes https://github.com/strimzi/strimzi-kafka-operator/issues/6058

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: operator-common/src/main/java/io/strimzi/operator/common/Operator.java [hunk 1]
```diff
@@ -36,6 +36,8 @@
  */
 public interface Operator {
 
+    static final ReconciliationLogger LOGGER = ReconciliationLogger.create(Operator.class);
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/Operator.java:39]: Why is this needed? You do not seem to use it.

File: operator-common/src/main/java/io/strimzi/operator/common/Operator.java [hunk 2]
```diff
@@ -71,16 +72,19 @@ default void reconcileAll(String trigger, String namespace, Handler<AsyncResult<
...
 
             for (NamespaceAndName resourceRef : desiredNames) {
-                resourceCounter(resourceRef.getNamespace()).getAndIncrement();
                 Reconciliation reconciliation = new Reconciliation(trigger, kind(), resourceRef.getNamespace(), resourceRef.getName());
                 futures.add(reconcile(reconciliation));
             }
             CompositeFuture.join(futures).map((Void) null).onComplete(handler);
         } else {
             resourceCounter(namespace).set(0);
+            pausedResourceCounter(namespace).set(0);
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/Operator.java:87]: How does this work with operator watching all namespaces? Line 76-77 will reset it for namespaces which have some resources. But these two lines would reset it for namespace `*` I suspect.

**Review by scholzj — COMMENTED** [review]
TBH, I'm not sure this really fixes it. I would expect that you will have to stream through the metrics map and reset it for each namespace that way.

File: operator-common/src/main/java/io/strimzi/operator/common/AbstractOperator.java [hunk 3]
```diff
@@ -175,6 +175,7 @@ public String kind() {
                     return Future.succeededFuture();
                 }
 
+                resourceCounter(namespace).getAndIncrement();
```
[the remark this answers is not in the document]
Comment (sknot-rh) [reply operator-common/src/main/java/io/strimzi/operator/common/AbstractOperator.java:178]: Ah, I did not know the metric should not be updated when it is created/removed.

[the remark this answers is not in the document]
Comment (scholzj) [reply operator-common/src/main/java/io/strimzi/operator/common/AbstractOperator.java:178]: The problem is that you cannot count it. You might get modified multiple times between the timer based reconciliations.

File: operator-common/src/main/java/io/strimzi/operator/common/Operator.java (hunk 2 above)
Comment (sknot-rh) [reply operator-common/src/main/java/io/strimzi/operator/common/Operator.java:87]: Correct

File: operator-common/src/main/java/io/strimzi/operator/common/Operator.java [hunk 4]
```diff
@@ -81,6 +86,7 @@ default void reconcileThese(String trigger, Set<NamespaceAndName> desiredNames,
...
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/Operator.java:89]: when running cluster wide operator, this would create metrics with `namespace=*`. I do not think we want that. I guess the question is whether we want to emit the metrics of 0 resources (since I guess no metric would also count as 0). And if we do, we should again if-it to set this only when it doesn't watch all namespaces.

Comment (scholzj) [reply operator-common/src/main/java/io/strimzi/operator/common/Operator.java:89]: Actually, these two lines 88+89 are not needed at all, or? They just set the 0 again already set on 72-77.

**Review by scholzj — APPROVED** [review]

**Review by ppatierno — APPROVED** [review]
LGTM
